src/cli_args.zig

```diff
@@ -0,0 +1,851 @@
+
+    pub fn deinit(self: CliArgs, gpa: mem.Allocator) void {
+        switch (self) {
+            .format => |fmt| gpa.free(fmt.paths),
+            .run => |run| gpa.free(run.app_args),
+            else => return,
+        }
+    }
+};
+
+/// Errors that can occur due to bad input while parsing the arguments
+pub const CliProblem = union(enum) {
+    missing_flag_value: struct {
+        flag: []const u8,
+    },
+    unexpected_argument: struct { cmd: []const u8, arg: []const u8 },
+    invalid_flag_value: struct {
+        value: []const u8,
+        flag: []const u8,
+        valid_options: []const u8,
+    },
+};
+
+/// The optimization strategy for the compilation of a Roc program
+pub const OptLevel = enum {
+    size, // binary size
+    speed, // binary speed
+    dev, // speed of compilation
+
+    fn from_str(str: []const u8) ?OptLevel {
+        if (mem.eql(u8, str, "speed")) return .speed;
+        if (mem.eql(u8, str, "size")) return .size;
+        if (mem.eql(u8, str, "dev")) return .dev;
+        return null;
+    }
+};
+
+/// Arguments for the default `roc` command
+pub const RunArgs = struct {
+    path: []const u8,
+    opt: OptLevel = .dev,
+    app_args: []const []const u8 = &[_][]const u8{},
+};
+
+/// Arguments for `roc check`
+pub const CheckArgs = struct {
+    path: []const u8,
+    main: ?[]const u8,
+};
+
+/// Arguments for `roc build`
+pub const BuildArgs = struct {
+    path: []const u8,
+    opt: OptLevel,
+    output: ?[]const u8 = null,
+};
+
+/// Arguments for `roc test`
+pub const TestArgs = struct { path: []const u8, opt: OptLevel, main: ?[]const u8 };
+
+/// Arguments for `roc format`
+pub const FormatArgs = struct {
+    paths: []const []const u8,
+    stdin: bool = false,
+    check: bool = false,
+};
+
+/// Arguments for `roc docs`
+pub const DocsArgs = struct {
+    path: []const u8,
+    output: []const u8,
+    root_dir: ?[]const u8 = null,
+};
+
+/// Parse a list of arguments.
+pub fn parse(gpa: mem.Allocator, args: []const []const u8) CliArgs {
+    if (args.len == 0) return parse_run(gpa, args);
+
+    if (mem.eql(u8, args[0], "check")) return parse_check(args[1..]);
+    if (mem.eql(u8, args[0], "build")) return parse_build(args[1..]);
+    if (mem.eql(u8, args[0], "format")) return parse_format(gpa, args[1..]);
+    if (mem.eql(u8, args[0], "test")) return parse_test(args[1..]);
+    if (mem.eql(u8, args[0], "repl")) return parse_repl(args[1..]);
+    if (mem.eql(u8, args[0], "version")) return parse_version(args[1..]);
+    if (mem.eql(u8, args[0], "docs")) return parse_docs(args[1..]);
+    if (mem.eql(u8, args[0], "help")) return CliArgs{ .help = main_help };
+    if (mem.eql(u8, args[0], "licenses")) return parse_licenses(args[1..]);
+
+    return parse_run(gpa, args);
+}
+
+const main_help =
+    \\Run the given .roc file
+    \\You can use one of the SUBCOMMANDS below to do something else!
+    \\
+    \\Usage: roc [OPTIONS] [ROC_FILE] [ARGS_FOR_APP]...
+    \\       roc <COMMAND>
```

Comment:
We call it a COMMAND here and a SUBCOMMAND above, also should this include the OPTIONS also?

`roc [SUBCOMMAND] [OPTIONS]`?